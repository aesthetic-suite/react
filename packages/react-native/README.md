# Aesthetic - React Native Integration

[![Build Status](https://github.com/aesthetic-suite/react/workflows/Build/badge.svg)](https://github.com/aesthetic-suite/react/actions?query=branch%3Amaster)
[![npm version](https://badge.fury.io/js/%40aesthetic%react-native.svg)](https://www.npmjs.com/package/@aesthetic/react-native)
[![npm deps](https://david-dm.org/aesthetic-suite/react.svg?path=packages/react-native)](https://www.npmjs.com/package/@aesthetic/react-native)

Provides styles for React Native components through hook or HOC based APIs. Built around the
powerful [Aesthetic](https://github.com/aesthetic-suite/framework) core library.

```tsx
import React from 'react';
import { Pressable, PressableProps } from 'react-native';
import { createComponentStyles, useStyles } from '@aesthetic/react-native';

export const styleSheet = createComponentStyles((theme) => ({
	button: {
		flex: 1,
		flexWrap: 'nowrap',
		alignItems: 'center',
		width: 'auto',
		padding: theme.tokens.spacing.df,
	},

	button_block: {
		width: '100%',
	},
}));

export interface ButtonProps extends PressableProps {
	children: React.ReactNode;
	block?: boolean;
}

export default function Button({ children, block = false, ...props }: ButtonProps) {
	const sx = useStyles(styleSheet);

	return (
		<Pressable {...props} style={sx('button', block && 'button_block')}>
			{children}
		</Pressable>
	);
}
```

## Features

- Hook and HOC based APIs for styling components, accessing themes, and handling directionality.
- Element level style sheets and nested themes.
- First-class directionality support (RTL, LTR).

## Requirements

- React 17+
- React Native 0.64+

## Installation

```
yarn add @aesthetic/react-native react react-native
```

## Documentation

[https://aestheticsuite.dev/docs/react-native](https://aestheticsuite.dev/docs/react-native)
