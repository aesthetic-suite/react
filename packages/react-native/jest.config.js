module.exports = {
  preset: 'react-native',
  setupFilesAfterEnv: ['../../tests/setup.ts'],
};
