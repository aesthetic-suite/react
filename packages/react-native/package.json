{
  "name": "@aesthetic/react-native",
  "version": "0.0.0",
  "description": "React Native hooks, composers, and more for styling components with Aesthetic.",
  "keywords": [
    "aesthetic",
    "react",
    "native",
    "styles",
    "hook",
    "hoc",
    "composer"
  ],
  "repository": {
    "type": "git",
    "url": "git@github.com:aesthetic-suite/react.git",
    "directory": "packages/react-native"
  },
  "author": "Miles Johnson",
  "license": "MIT",
  "main": "./lib/index.js",
  "module": "./esm/index.js",
  "types": "./dts/index.d.ts",
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@aesthetic/core": "^0.9.0",
    "@aesthetic/core-react": "^0.1.5",
    "@aesthetic/utils": "^0.7.0"
  },
  "peerDependencies": {
    "react": "^17.0.0",
    "react-native": "^0.64.0"
  },
  "devDependencies": {
    "@testing-library/react-native": "^7.0.2",
    "@types/react-native": "^0.64.4",
    "react": "^17.0.1",
    "react-native": "^0.64.0"
  },
  "funding": {
    "type": "ko-fi",
    "url": "https://ko-fi.com/milesjohnson"
  },
  "packemon": {
    "platform": "native"
  },
  "tsconfig": {
    "compilerOptions": {
      "jsx": "react",
      "lib": [
        "esnext"
      ],
      "types": [
        "node",
        "react",
        "react-native"
      ]
    }
  }
}
