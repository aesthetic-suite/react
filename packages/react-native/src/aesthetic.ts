import { Appearance, I18nManager } from 'react-native';
import { Aesthetic, RenderResult, RenderResultVariant, Rule } from '@aesthetic/core';
import { objectLoop } from '@aesthetic/utils';
import { NativeEngine, NativeRule, NativeStyles } from './types';

const noop = () => {};

function unsupported(method: string) {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (): any => {
		// eslint-disable-next-line no-console
		console.warn(`${method} is not supported by React Native.`);
	};
}

function renderRule(rule: Rule): RenderResult<NativeStyles> {
	const { '@variants': variantStyles, ...result } = rule as NativeRule;
	const variants: RenderResultVariant<NativeStyles>[] = [];

	if (variantStyles) {
		objectLoop(variantStyles, (variant, type) => {
			variants.push({
				result: variant,
				types: type.split('+').map((v) => v.trim()),
			});
		});
	}

	return { result, variants };
}

const engine: NativeEngine = {
	atomic: false,
	direction: I18nManager.isRTL ? 'rtl' : 'ltr',
	prefersColorScheme: (scheme) => scheme === Appearance.getColorScheme(),
	prefersContrastLevel: () => false,
	renderDeclaration: unsupported('renderDeclaration()'),
	renderFontFace: unsupported('renderFontFace()'),
	renderImport: unsupported('renderImport()'),
	renderKeyframes: unsupported('renderKeyframes()'),
	renderRule,
	renderRuleGrouped: unsupported('renderRuleGrouped()'),
	renderVariable: unsupported('renderVariable()'),
	ruleIndex: 0,
	setDirection: noop,
	setRootVariables: noop,
	setTheme: noop,
};

export const aesthetic = new Aesthetic<NativeStyles, NativeRule>();

aesthetic.configureEngine(engine);
