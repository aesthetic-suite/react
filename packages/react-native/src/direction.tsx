import React from 'react';
import { View } from 'react-native';
import { createDirectionHelpers } from '@aesthetic/core-react';
import { aesthetic } from './aesthetic';

function DirectionWrapper({ children, ...props }: { children: React.ReactNode }) {
	return (
		<View testID="aesthetic-direction-provider" {...props}>
			{children}
		</View>
	);
}

export const {
	DirectionContext,
	DirectionProvider,
	useDirection,
	withDirection,
} = createDirectionHelpers(aesthetic, { wrapper: DirectionWrapper });
