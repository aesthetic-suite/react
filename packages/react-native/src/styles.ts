import React, { createElement, forwardRef, memo, useMemo } from 'react';
import { Utilities } from '@aesthetic/core';
import { createStyleHelpers } from '@aesthetic/core-react';
import { arrayLoop, toArray } from '@aesthetic/utils';
import { aesthetic } from './aesthetic';
import { useDirection } from './direction';
import { useTheme } from './theme';
import { InferProps, NativeBlock, NativeStyles, StyledComponent } from './types';

export const { getVariantsFromProps, useStyles, withStyles } = createStyleHelpers<
	NativeStyles,
	NativeBlock,
	NativeStyles[]
>(aesthetic, {
	generate(args, variants, results) {
		const style: NativeStyles[] = [];

		arrayLoop(args, (arg) => {
			if (!arg) {
				return;
			}

			if (Array.isArray(arg)) {
				arg.forEach((a) => {
					if (a) {
						style.push(a);
					}
				});

				return;
			}

			const hash = results[arg];

			if (!hash) {
				return;
			}

			if (hash.result) {
				style.push(hash.result);
			}

			if (hash.variants) {
				arrayLoop(hash.variants, ({ types, result }) => {
					if (types.every((type) => variants.has(type))) {
						style.push(result);
					}
				});
			}
		});

		return style;
	},
	useDirection,
	useTheme,
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createStyled<T extends React.ComponentType<any>, V extends object = {}>(
	type: T,
	factory: NativeBlock | ((utilities: Utilities<NativeBlock>) => NativeBlock),
): StyledComponent<InferProps<T> & V> {
	if (process.env.NODE_ENV !== 'production') {
		const typeOfType = typeof type;
		const typeOfFactory = typeof factory;

		if (typeOfType !== 'string' && typeOfType !== 'function' && typeOfType !== 'object') {
			throw new TypeError(
				`Styled components must extend a View or React component, found ${typeOfType}.`,
			);
		}

		if (typeOfFactory !== 'function' && typeOfFactory !== 'object') {
			throw new TypeError(
				`Styled components require a style sheet factory function, found ${typeOfFactory}.`,
			);
		}
	}

	const styleSheet = aesthetic.createComponentStyles((utils) => ({
		element: typeof factory === 'function' ? factory(utils) : factory,
	}));

	const Component = memo(
		forwardRef((baseProps, ref) => {
			const sx = useStyles(styleSheet);
			const { props, variants } = getVariantsFromProps<'style'>(sx.result.element, baseProps);
			const style = variants ? sx(variants, 'element') : sx('element');

			// Cache the style prop so that we avoid unwanted rerenders
			const styleProp = useMemo(() => {
				const viewStyles = [...style];

				if (props.style) {
					viewStyles.push(...toArray(props.style));
				}

				return viewStyles;
			}, [style, props.style]);

			return createElement(type, {
				...props,
				ref,
				style: styleProp,
			});
		}),
	);

	const displayName =
		typeof type === 'string'
			? String(type)
			: String((type as React.ComponentType).displayName ?? (type as Function).name);

	Component.displayName = `styled(${displayName})`;

	(Component as StyledComponent).styleSheet = styleSheet;

	// Use the return type of `createStyled` instead of `forwardRef`
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-explicit-any
	return Component as any;
}
