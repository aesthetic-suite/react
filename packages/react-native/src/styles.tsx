import React from 'react';
import { Utilities } from '@aesthetic/core';
import { createStyleHelpers } from '@aesthetic/core-react';
import { arrayLoop } from '@aesthetic/utils';
import aesthetic from './aesthetic';
import { useDirection } from './direction';
import { useTheme } from './theme';
import { InferProps, StyledComponent, NativeBlock, NativeStyles } from './types';

export const { getVariantsFromProps, useStyles, withStyles } = createStyleHelpers(aesthetic, {
  generate(keys, variants, results) {
    let style: NativeStyles = {};

    arrayLoop(keys, (key) => {
      const hash = results[key];

      if (!hash) {
        return;
      }

      if (hash.result) {
        style = hash.result;
      }

      if (hash.variants) {
        arrayLoop(variants, (variant) => {
          if (hash.variants?.[variant]) {
            Object.assign(style, hash.variants[variant]);
          }
        });
      }
    });

    return style;
  },
  useDirection,
  useTheme,
});

export function createStyled<T extends React.ComponentType<any>, V extends object = {}>(
  type: T,
  factory: NativeBlock | ((utilities: Utilities<NativeBlock>) => NativeBlock),
): StyledComponent<InferProps<T> & V> {
  if (__DEV__) {
    const typeOfType = typeof type;
    const typeOfFactory = typeof factory;

    if (typeOfType !== 'string' && typeOfType !== 'function' && typeOfType !== 'object') {
      throw new TypeError(`Styled components must extend a View component, found ${typeOfType}.`);
    }

    if (typeOfFactory !== 'function' && typeOfFactory !== 'object') {
      throw new TypeError(
        `Styled components require a style sheet factory function, found ${typeOfFactory}.`,
      );
    }
  }

  const styleSheet = aesthetic.createComponentStyles((utils) => ({
    element: typeof factory === 'function' ? factory(utils) : factory,
  }));

  const Component = React.memo(
    React.forwardRef((baseProps, ref) => {
      const cx = useStyles(styleSheet);
      const { props, variants } = getVariantsFromProps<'style'>(styleSheet, baseProps);
      let style: NativeStyles | NativeStyles[] = cx(variants, 'element');

      if (props.style) {
        style = [style, props.style];
      }

      return React.createElement(type, {
        ...props,
        ref,
        style,
      });
    }),
  );

  const displayName =
    typeof type === 'string'
      ? String(type)
      : String((type as React.ComponentType).displayName || (type as Function).name);

  Component.displayName = `styled(${displayName})`;

  (Component as StyledComponent).styleSheet = styleSheet;

  // Use the return type of `createStyled` instead of `forwardRef`
  return Component as any;
}
