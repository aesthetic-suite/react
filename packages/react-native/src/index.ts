/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import aesthetic from './aesthetic';
import { createStyled } from './styles';

export * from './direction';
export * from './styles';
export * from './theme';
export * from './types';
export * from '@aesthetic/core';
export * from '@aesthetic/core-react';

export const {
  changeDirection,
  changeTheme,
  configure,
  createComponentStyles,
  createElementStyles,
  getActiveDirection,
  getActiveTheme,
  getEngine,
  getTheme,
  registerDefaultTheme,
  registerTheme,
  renderComponentStyles,
  subscribe,
  unsubscribe,
} = aesthetic;

// Export aliases
export const style = createElementStyles;
export const styles = createComponentStyles;
export const styled = createStyled;
