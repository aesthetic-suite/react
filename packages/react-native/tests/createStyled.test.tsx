/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { Text, TextInput, View, ViewStyle } from 'react-native';
import { Sheet } from '@aesthetic/core';
import { render } from '@testing-library/react-native';
import { createStyled } from '../src';
import { Wrapper } from './__fixtures__/Button';
import { setupAestheticReact, teardownAestheticReact } from './helpers';

function extractDebug(debug: () => void): string {
	const spy = jest.spyOn(console, 'log').mockImplementationOnce(() => {});

	debug();

	const output = spy.mock.calls[0][0];

	spy.mockRestore();

	return output;
}

describe('createStyled()', () => {
	beforeEach(() => {
		setupAestheticReact();
	});

	it('errors for a non-react element type', () => {
		expect(() =>
			createStyled(
				// @ts-expect-error Invalid type
				123,
				{},
			),
		).toThrow('Styled components must extend a View or React component, found number.');
	});

	it('errors for invalid style sheet factory', () => {
		expect(() =>
			createStyled(
				View,
				// @ts-expect-error Invalid type
				true,
			),
		).toThrow('Styled components require a style sheet factory function, found boolean.');
	});

	it('sets static properties on component', () => {
		const Button = createStyled(View, {});

		expect(Button.displayName).toBe('styled(View)');
		expect(Button.styleSheet).toBeInstanceOf(Sheet);

		const ComposedButton = createStyled(Button, {});

		expect(ComposedButton.displayName).toBe('styled(styled(View))');
		expect(ComposedButton.styleSheet).toBeInstanceOf(Sheet);
	});

	it('creates and renders a button with defined styles', () => {
		const Button = createStyled(View, () => ({
			display: 'flex',
			textAlign: 'center',
			padding: '1rem',
		}));

		const { debug } = render(
			<Button>
				<Text>Test</Text>
			</Button>,
			{
				wrapper: Wrapper,
			},
		);

		expect(extractDebug(debug)).toMatchSnapshot();
	});

	it('only renders styles once (create and mount), even when component rerenders', () => {
		const spy = jest.fn(
			() =>
				({
					fontWeight: 'bold',
				} as const),
		);

		const Link = createStyled(Text, spy);
		const cb = () => {};

		const { debug, update } = render(<Link onPress={cb}>Test</Link>, {
			wrapper: Wrapper,
		});

		update(<Link onPress={cb}>Test</Link>);
		update(<Link onPress={cb}>Test</Link>);
		update(<Link onPress={cb}>Test</Link>);

		expect(spy).toHaveBeenCalledTimes(1);
		expect(extractDebug(debug)).toMatchSnapshot();
	});

	it('can pass custom props/attributes', () => {
		const Input = createStyled(TextInput, {
			borderWidth: 1,
			borderColor: 'black',
			padding: '1rem',
		});

		const { debug } = render(
			<Input editable={false} placeholder="Search..." style={{ borderStyle: 'solid' }} />,
			{
				wrapper: Wrapper,
			},
		);

		expect(extractDebug(debug)).toMatchSnapshot();
	});

	describe('variants', () => {
		interface AlertProps {
			palette?: 'negative' | 'positive' | 'warning';
		}

		function createAlert() {
			return createStyled<typeof View, AlertProps>(View, (css) => ({
				backgroundColor: css.tokens.palette.neutral.bg.base,
				'@variants': {
					'palette:success': {
						backgroundColor: css.tokens.palette.positive.bg.base,
					},
					'palette:failure': {
						backgroundColor: css.tokens.palette.negative.bg.base,
					},
					'palette:warning': {
						backgroundColor: css.tokens.palette.warning.bg.base,
					},
				},
			}));
		}

		it('supports variants', () => {
			const Alert = createAlert();
			const children = <Text>Title and other content!</Text>;

			const { debug, update } = render(<Alert palette="positive">{children}</Alert>, {
				wrapper: Wrapper,
			});

			expect(extractDebug(debug)).toMatchSnapshot();

			update(<Alert palette="negative">{children}</Alert>);

			expect(extractDebug(debug)).toMatchSnapshot();

			update(<Alert>{children}</Alert>);

			expect(extractDebug(debug)).toMatchSnapshot();
		});

		it('doesnt pass the variant prop to the View element', () => {
			const Alert = createAlert();

			const { debug } = render(<Alert palette="positive" />, {
				wrapper: Wrapper,
			});

			expect(extractDebug(debug)).toMatchSnapshot();
		});

		it('inherits and merges variant props and types when composing', () => {
			const Alert = createAlert();

			const SubAlert = createStyled<typeof Alert, { size?: 'lg' | 'sm' }>(Alert, {
				'@variants': {
					'size:sm': { padding: 1 },
					'size:lg': { padding: 2 },
				},
			});

			const { debug } = render(<SubAlert palette="positive" size="lg" />, {
				wrapper: Wrapper,
			});

			expect(extractDebug(debug)).toMatchSnapshot();
		});
	});

	describe('composition', () => {
		it('supports extending non-styled components', () => {
			function Base({ children, style }: { children: React.ReactNode; style?: ViewStyle }) {
				return <View style={style}>{children}</View>;
			}

			const Button = createStyled(Base, () => ({
				display: 'flex',
				textAlign: 'center',
				padding: '1rem',
			}));

			const { debug } = render(<Button style={{ color: 'red' }}>Normal</Button>, {
				wrapper: Wrapper,
			});

			expect(extractDebug(debug)).toMatchSnapshot();
		});

		it('supports extending other styled components', () => {
			const Button = createStyled(View, () => ({
				display: 'flex',
				textAlign: 'center',
				padding: '1rem',
			}));

			const BlockButton = createStyled(Button, {
				display: 'flex',
				width: '100%',
			});

			const LargeBlockButton = createStyled(BlockButton, {
				padding: '2rem',
				fontSize: 18,
			});

			const { debug } = render(
				<>
					<Button>Normal</Button>
					<BlockButton>Block</BlockButton>
					<LargeBlockButton>Large</LargeBlockButton>
				</>,
				{
					wrapper: Wrapper,
				},
			);

			expect(extractDebug(debug)).toMatchSnapshot();
		});

		it('can access the ref from multiple layers of composition', () => {
			const Leaf = createStyled(View, {});
			const Branch = createStyled(Leaf, {});
			const Trunk = createStyled(Branch, {});
			const Root = createStyled(Trunk, {});

			const ref = { ref: true };
			const spy = jest.fn();
			// @ts-expect-error Allow refs
			const { debug } = render(<Root ref={spy} />, {
				createNodeMock: () => ref,
				wrapper: Wrapper,
			});

			expect(spy).toHaveBeenCalled();
			expect(extractDebug(debug)).toMatchSnapshot();
		});
	});
});
