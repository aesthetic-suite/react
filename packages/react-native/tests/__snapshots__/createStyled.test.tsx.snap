// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyled() can pass custom props/attributes 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<TextInput[39m
    [33mallowFontScaling[39m=[32m{true}[39m
    [33meditable[39m=[32m{false}[39m
    [33mplaceholder[39m=[32m\\"Search...\\"[39m
    [33mrejectResponderTermination[39m=[32m{true}[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"borderColor\\": \\"black\\",
          \\"borderWidth\\": 1,
          \\"padding\\": \\"1rem\\",
        },
        Object {
          \\"borderStyle\\": \\"solid\\",
        },
      ]
    }[39m
    [33munderlineColorAndroid[39m=[32m\\"transparent\\"[39m
  [36m/>[39m
[36m</View>[39m"
`;

exports[`createStyled() composition can access the ref from multiple layers of composition 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {},
        Object {},
        Object {},
        Object {},
      ]
    }[39m
  [36m/>[39m
[36m</View>[39m"
`;

exports[`createStyled() composition supports extending non-styled components 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"padding\\": \\"1rem\\",
          \\"textAlign\\": \\"center\\",
        },
        Object {
          \\"color\\": \\"red\\",
        },
      ]
    }[39m
  [36m>[39m
    [0mNormal[0m
  [36m</View>[39m
[36m</View>[39m"
`;

exports[`createStyled() composition supports extending other styled components 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"padding\\": \\"1rem\\",
          \\"textAlign\\": \\"center\\",
        },
      ]
    }[39m
  [36m>[39m
    [0mNormal[0m
  [36m</View>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"padding\\": \\"1rem\\",
          \\"textAlign\\": \\"center\\",
        },
        Object {
          \\"display\\": \\"flex\\",
          \\"width\\": \\"100%\\",
        },
      ]
    }[39m
  [36m>[39m
    [0mBlock[0m
  [36m</View>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"padding\\": \\"1rem\\",
          \\"textAlign\\": \\"center\\",
        },
        Object {
          \\"display\\": \\"flex\\",
          \\"width\\": \\"100%\\",
        },
        Object {
          \\"fontSize\\": 18,
          \\"padding\\": \\"2rem\\",
        },
      ]
    }[39m
  [36m>[39m
    [0mLarge[0m
  [36m</View>[39m
[36m</View>[39m"
`;

exports[`createStyled() creates and renders a button with defined styles 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"padding\\": \\"1rem\\",
          \\"textAlign\\": \\"center\\",
        },
      ]
    }[39m
  [36m>[39m
    [36m<Text>[39m
      [0mTest[0m
    [36m</Text>[39m
  [36m</View>[39m
[36m</View>[39m"
`;

exports[`createStyled() only renders styles twice (create and mount), even when component rerenders 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<Text[39m
    [33monPress[39m=[32m{[Function cb]}[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"fontWeight\\": \\"bold\\",
        },
      ]
    }[39m
  [36m>[39m
    [0mTest[0m
  [36m</Text>[39m
[36m</View>[39m"
`;

exports[`createStyled() variants doesnt pass the variant prop to the View element 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"backgroundColor\\": \\"#000\\",
        },
      ]
    }[39m
  [36m/>[39m
[36m</View>[39m"
`;

exports[`createStyled() variants inherits and merges variant props and types when composing 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"backgroundColor\\": \\"#000\\",
        },
        Object {},
        Object {
          \\"padding\\": 2,
        },
      ]
    }[39m
  [36m/>[39m
[36m</View>[39m"
`;

exports[`createStyled() variants supports variants 1`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"backgroundColor\\": \\"#000\\",
        },
      ]
    }[39m
  [36m>[39m
    [36m<Text>[39m
      [0mTitle and other content![0m
    [36m</Text>[39m
  [36m</View>[39m
[36m</View>[39m"
`;

exports[`createStyled() variants supports variants 2`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"backgroundColor\\": \\"#000\\",
        },
      ]
    }[39m
  [36m>[39m
    [36m<Text>[39m
      [0mTitle and other content![0m
    [36m</Text>[39m
  [36m</View>[39m
[36m</View>[39m"
`;

exports[`createStyled() variants supports variants 3`] = `
"[36m<View[39m
  [33mdir[39m=[32m\\"ltr\\"[39m
[36m>[39m
  [36m<View[39m
    [33mstyle[39m=[32m{
      Array [
        Object {
          \\"backgroundColor\\": \\"#000\\",
        },
      ]
    }[39m
  [36m>[39m
    [36m<Text>[39m
      [0mTitle and other content![0m
    [36m</Text>[39m
  [36m</View>[39m
[36m</View>[39m"
`;
